### ISSUE TITLE
Android app creates new SSE subsciptions on every page change

### ISSUE THREAD

**Issue by kaikreuzer** [openhab/openhab-android#2215]
***NOTE:*** This has first been posted in openhab-core with the below description and discussion. With https://github.com/openhab/openhab-android/issues/2215#issuecomment-698545471 it turns out that root cause is in the Android app.

--------------------------

When UIs try to subscribe to sitemap events through `/rest/sitemaps/events/subscribe`, I am seeing HTTP 503 being returned with body
```
{
  "error": {
    "message": "Max number of subscriptions is reached.",
    "http-code": 503
  }
}
```
It seems that there is some regression as I am only using 1 or two clients at a time.

**Comment by maniac103** [comment]
I wouldn't consider keeping more than one open SSE connection a bug. The app is currently designed to keep pages in the back stack updated, so it opens one subscription for each page.
Since it seems unlikely that the back stacks grows to more than 50 pages, I think the *actual* problem is the subscriptions not being released properly when navigating back. Is there anything we need to do to release a subscription besides closing the connection?

**Comment by wborn** [comment]
It would be interesting to know if this issue also occurs with 2.5.x. If so it would most likely be an issue in the App. In OH3 there is a job to clean dead subscriptions every 5 minutes, see the [SitemapResource](https://github.com/openhab/openhab-core/blob/369e678a6347f1ab41fdd8eb822604c3c2a3c923/bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java#L189-L198) which could also have an issue.

**Comment by maniac103** [comment]
OK, I checked the code a bit. To recap, the app currently for every new page does
- create a new subscription
- creates an event stream using that subscription and page ID of the new page

What I think might work from my understanding of the SSE API is
- have a pool of previously used subscriptions
- when opening a new page, try to reuse a pooled subscription
- if event stream creation yields a 404 (subscription discarded already) or other error, remove it from pool
- if no pooled subscription exists, create a new one

Does that make sense to everyone? I think that approach would limit the number of subscriptions to 'number of open pages + pool size'.

**Comment by wborn** [comment]
I agree that using a pool would be a nice approach to limit the number of subscriptions. How many pages does it keep open in the background? There could still be issues for users who run the App on many devices.

**Comment by maniac103** [comment]
> How many pages does it keep open in the background?

Depends on navigation depth. There's no limit coded into the app itself, so the limit is the nesting depth in the respective sitemap.

**Comment by kaikreuzer** [comment]
> Depends on navigation depth.

So does this mean that it only creates new subscriptions when navigating down to a page, but not when navigating to another page on the same hierarchy level? In this case, I'd agree that it should be fine to keep one subscription open for every depth level. But this is not what I see in the logs - it seems that a new subscription is invariably created on every single page change, which definitely is not good (and I do not see any subscriptions being release either).

> In OH3 there is a job to clean dead subscriptions every 5 minutes, see the SitemapResource which could also have an issue.

This job seems to indeed have an issue. I see an every increasing list of open subscriptions and the cleaning of stale subscriptions after 5 minutes simply does not seem to work - so that's something we'll indeed have to address in the core code in addition. I have created https://github.com/openhab/openhab-core/issues/1674 to track that.

**Comment by maniac103** [comment]
> But this is not what I see in the logs - it seems that a new subscription is invariably created on every single page change, which definitely is not good

Yes, that's the case currently. I described the state after the change, in which case there still would be one open subscription per page in the back stack.

> (and I do not see any subscriptions being release either).

How can a client release subscriptions? I asked earlier already because I don't see a way of doing that.

**Comment by wborn** [comment]
> How can a client release subscriptions? I asked earlier already because I don't see a way of doing that.

When the client closes the connection this should be detected by the server (https://github.com/openhab/openhab-android/issues/2215#issuecomment-699495454) and it will remove the subscription. The server has to deal with clients that  can disconnect at any time anyway so that probably explains why there is no `unsubscribe` : 

![Screenshot from 2020-09-28 09-43-52](https://user-images.githubusercontent.com/12213581/94406418-49bdb100-0172-11eb-807e-c1d9da90a93c.png)

**Comment by maniac103** [comment]
> When the client closes the connection this should be detected by the server (#2215 (comment)) and it will remove the subscription. 

But - if I understand that comment correctly - it will be discarded only after 5 minutes, right? Which would mean there's no way for the client to relase subscriptions immediately? If that's the case, the subscription pooling is the only solution I can see.

**Comment by wborn** [comment]
Yes the pooling will certainly help with reducing the number of active subscriptions as will fixing any cleanup bugs and perhaps it should also run the cleanup more often and maybe there is also a way to instantly handle disconnecting clients. But I can imagine the current cleanup logic was introduced to prevent concurrency issues when subscriptions are being reused.

**Comment by kaikreuzer** [comment]
> so that probably explains why there is no unsubscribe

I was actually expecting that we also have an `unsubscribe` but this is missing indeed. I think it would not harm to add it as it would help to cleanly manage the subscriptions in situations as the Android app, where multiple subscriptions are used at the same time. Wdyt, @wborn?

**Comment by wborn** [comment]
After some debugging and fixes I think the cleanup works a lot better with https://github.com/openhab/openhab-core/pull/1679.

**Comment by mueller-ma** [comment]
> I was actually expecting that we also have an unsubscribe but this is missing indeed.

Was there any progress on this?

**Comment by wborn** [comment]
The [SitemapResource history](https://github.com/openhab/openhab-core/commits/main/bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/internal/SitemapResource.java) doesn't suggest an unsubscribe operation was added.
So #1679 fixed/prevented most the issues users have.

**Comment by maniac103** [comment]
@wborn So from the data we have, the pooling solution I suggested above is no longer necessary?

**Comment by wborn** [comment]
It's probably nice to have now. Just like having an unsubscribe method. :wink:
